Start fixing module

Micky1979 opened the pull request:
Hi meklort. I want to try the module with the updated kext but there are some to change. Please take a look.

Micky1979 committed:
Merge remote-tracking branch 'meklort/master'

Micky1979 committed:
Start fixing module

xml.c is buggie reading data tag. Introduced a helper function to clean
end of line and tab characters. Also adding bytes instead of a base 64
string to the nvram node
(need to test the module with the new kext, but some problems to fix)

Micky1979 committed:
fixed misspelling

Micky1979 committed:
fixed compilation

added a cast to const uint8_t * for getSmbiosUUID()

Micky1979 committed:
fixed style

fixed style to match the rest of the code

Micky1979 committed:
Updated module version to 6.9.5

Updated module version to 6.9.5, will win against older one if already
present somewhere.

Micky1979 committed:
Including new header in the Xcode project

Version.h included in the Xcode project, fixed init message output

Micky1979 committed:
init message fixed

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -103,6 +103,9 @@ bool FileNVRAM::start(IOService *provider)
     setPath(OSString::withCString(FILE_NVRAM_PATH));
     
+    // register immediately, save a lot of startup time
+    registerNVRAM();
```
Please do *not* register the nvram module untill after the contents are valid.

For booting w/ a supported boot loader, the means you can do it after the properties are copied.
If the NVRAM is read off of disk, that means it has to be after the callback.

I'd suggest creating a new function, say
FileNVRAM::finalizeInit()
{
    // set mInitComplete = true;
    // registerNVRAM with platform expert
    // other?
}

Any time the code currently sets mInitComplete = true, instead, call the new function.

Micky1979 replied:
I'll do that!
PS tomorow, after lesson and job

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -115,49 +118,97 @@ bool FileNVRAM::start(IOService *provider)
-        mInitComplete = true;
-        mSafeToSync = true;
-        registerNVRAM();
+        mInitComplete = true;        mSafeToSync = true;
```
Use separate lines.

Micky1979 replied:
Sorry was unwanted

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -115,49 +118,97 @@ bool FileNVRAM::start(IOService *provider)
+    notifier = NULL;
     
-    return true;
+    if(OSDictionary *matching = nameMatching("Apple_HFS"))
```
Recent version of XNU (finally) support more than just hfs+ for the rootfs. For example, zfs can now be used.

Why is this code matching on Apple_HFS?

It also looks like this code never actually verifies that the rootfs was mounted, and just assumes that if one IOMediaBSDClient (HFS) class is found, that the rootfs must have been mounted.

Looking at the ioregistery on my mac (with an encrypted rootfs), I see that the EFI partition shows up as an IOMediaBSDClient even though it has *not* mounted. Yes, the EFI partition probably won't be HFS, but it's entirely possible for other HFS partitions to exist on other driver (or none to exist at all).

So... it looks like there may be a race condition with how this code is implemented

*Note: there is an actual rootfs mounted callback in XNU, however they way Apple implemented it, only one kext can reliably register for it (and I think it's a private API). In any case, it looks like more work needs to be done here to ensure everything always works.

Micky1979 replied:
zfs? as a third part kernel extension?
There is something about "mpo_mount_check_mount" that can be used in a kernel extensions, but I need to study it. Let you know.

meklort replied:
Yes, the zfs kext has support for using a zfs file system as the rootfs.
Recent versions of XNU exposed the apis for it. Older versions of XNU required a bit of trickery to get it to work right.

meklort commented on module/FileNVRAM.c:
```diff
@@ -359,7 +388,18 @@ static void readplist()
+                    // but I want it to be: <data>fw==</data>
+                    //
+                    // so some garbage ends up in /chosen/nvram. just remove them!
+                    XMLParseFile(strdelchar(strdelchar(plistBase, '\n'), '\t'), &gPListData);
```
Is this here due to a bug in chameleon?
If so, I'd suggest noting that and preferably fixing chameleon. I don't mind having a workaround here, but please make sure chameleon is fixed as well.

Micky1979 replied:
I think is there for years. XMLCastData is only used by this module. Maybe I can do that here, but a (normal) plist file really contains end of line and tabs under the data tag.

meklort replied:
Yes, I agree that the plist has \n and \t's in it, and I'm fine with the code here (for compat with older chameleon versions), I'm just wondering if it's a bug in chameleon's XMLParseFile function. And if so, then that should be fixed as well.

Micky1979 replied:
Should I reimplement it? I don't had the time to look at chameleon, but I think that I can do that in the BASE64Decode() function (which in truth should fail if there are foreign characters). Or in ParseTagData()

Micky1979 commented:
Idea. OSKextRequestResource() can be used? If can read inside the Resources folder (of any other  extensions)..than we know that the root fs is ready. What do you think?

meklort commented:
From looking at the source code and documentation, OSKextRequestResource would probably work, however, this would wait for userland to start up before registering nvram. That might be later than desired and could slow down the boot process.

Micky1979 commented:
Got IOBSD. Service registered if we have datas from the bootloader, waiting for IOBSD to safe sync.
IOKitBSDInit has its own timeout and apparently we don't need another one here: got it or not.
Is that correct? 

EDIT
waiting for a client to appear through notification and than looks for IOBSD is really fast, no visible delay. The delay is present only if the bootloader has not provided the nvram datas because the service isn't registered, but also this way no longer hangs.

meklort commented:
Please be aware that when IOBSD is published in the IORegistery, the root file system is not yet mounted. I does get mounted very quickly after, but there is a race condition because of this.

The old code used a timer to poll for the file being available for a while to solve this.

The best way possible to ensure that the rootfs is mounted is to use the mountroot_post_hook callback in the kernel, however the way it's implemented does not allow multiple kexts to use it reliable (and apple already uses it), so instead, we must pull until the rootfs gets mounted. Please add this code back unless if you have a better way to work around the race condition.

meklort commented:
TO be clear, the new code does get called multiple times, however the class that the notification happens, while closer to the rootfs being mounted, can also happen before rootfs is mounter.

meklort commented:
Ok, I just checked the IORegistery, and it looks like the match might actually be correct. I'll have ot look at it a bit more to confirm, but, it looks like no change is needed for now.

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -115,49 +115,115 @@ bool FileNVRAM::start(IOService *provider)
+    FileNVRAM *self = (FileNVRAM *)target;
     
-    mInitComplete = true;
+    notifier->remove();
```
I'd suggest only removing the notifier once we've actually been able to read the rootfs. If we aren't able to read the rootfs the first time (say, because the actual device matched was something other than the root), we can try again on the next match.

Micky1979 replied:
In fact the notification was removed thinking that that was enough. May be that the first one is not what we are looking for.

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -115,49 +115,115 @@ bool FileNVRAM::start(IOService *provider)
+    notifier = NULL;
     
-    return true;
+    if(OSDictionary *matching = IOService::resourceMatching( "IOBSD" ))
```
Declare OSDictionary *matching outside of the if() block.

meklort replied:
I'm also not sure if this is doing what you think it's doing. I'm pretty sure this does *not* check to make sure the device that matched has an IOResourceMatch property of IOBSD. I'm pretty sure this will always enter the if block the kernel has free ram.

You probably want to check that the passed in service actually matches the IOBSD resource.

Micky1979 replied:
It should return something that match, but I'll try as you said.

meklort replied:
The IOService::resourceMatching function *only* allocates an OSDictionary that is supposed to be used as the argument to the addMatchingNotification function.

What you probably want to do instead, is to use service->getProperty("IOResourceMatch") and check to see if the result is equal to "IOBSD". 

The key things to note here are:
This asks the service instance something. The IOService::resourceMatching function has no relation to the service instance, and as a result cannot give you any useful information.
This grabs the contents of the "IOResourceMarch" property, which is what the regular matching notification function would check.

Micky1979 replied:
A question (hope to not bore you :-) ).
I'm testing, and I'm going to reintroduce the old code w/o making additional commits for now, but I wondering why to not set the FileNVRAM personality this way:

`<key>IOResourceMatch</key>
			<string>IOBSD</string>`

instead of efi-runtime. Any downside doing that?
Thanks.

meklort commented on kext/FileNVRAM/FileNVRAM.h:
```diff
@@ -180,7 +182,7 @@ class FileNVRAM : public IODTNVRAM
     IOCommandGate* mCommandGate;
     OSString*      mFilePath;
-    IOTimerEventSource* mTimer;
+    IONotifier* disk_notifier_start_;
```
This does not need to be a member variable.

Micky1979 replied:
Due to the fact that is reachable in the call back, agreed.

Micky1979 replied:
shoul not be:  `IOTimerEventSource* mTimer = IOTimerEventSource::timerEventSource(this, timeoutOccurred);`
just:
`mTimer = IOTimerEventSource::timerEventSource(this, timeoutOccurred);` ?

Micky1979 replied:
Ha ha, that was the problem now it works. So mutch trouble for a little typo :-). Not really visible ... I watched it multiple times but I realize that only now reintroducing the old code by copy/paste. I think is a big bug in clang to not showing any warning/error!

Micky1979 replied:
Should I revert (I mean delete) all my commits to that one: [da513ff3ce7b4a5cca427b5e352131f077f61edf](https://github.com/meklort/FileNVRAM/pull/3/commits/da513ff3ce7b4a5cca427b5e352131f077f61edf) ?
init message fixed, containing also the strdelchar() function...

meklort replied:
Good catch. 

Yes, If the existing code works with that fix, I'd try just making that (minimal) change.

DO note that while the code is valid c++, it is incorrect. I would not expect clang to catch it, however I *would* expect the clang static analyzer (a seperate executable) to catch it, or something like pclint / flexelint (I am planning to run flexelint on it at some point, when I get a chance).

Micky1979 replied:
I've just deleted some commits to have a more clear history. I would leave the changes for the FileNVRAM module there for now (my own purpose). Sir you don't need this pull request for that little change and we can close it, just apply by your self. Later, after testing with chameleon I will open a new pull request.... I have to test the internal kernel/kexts patcher using the module and ensure that no conflicts exists or make it conditional when the module is loaded. Thanks for the patience!

meklort commented on kext/FileNVRAM/FileNVRAM.cpp:
```diff
@@ -115,49 +115,115 @@ bool FileNVRAM::start(IOService *provider)
-        
-        if(mTimer)
+        // waiting for a client to appear before watching for IOBSD
+        if(VERSION_MAJOR >10)
```
I'm not sure if this is what you want. This is a compile time constant ad has no correlation to the actual kernel that you are running on.

Micky1979 replied:
? Well, then that should be totally useless. I'll look for real values.

meklort replied:
Please take a look at libkern/version.h for more information. It shows what you'll want to use.
You can find a copy here: /System/Library/Frameworks/Kernel.framework/Headers/libkern/version.h

Micky1979 replied:
Done. Tested using version_major instead of VERSION_MAJOR and it do the job.

Micky1979 commented on kext/FileNVRAM/FileNVRAM.h:
```diff
@@ -180,7 +182,7 @@ class FileNVRAM : public IODTNVRAM
     IOCommandGate* mCommandGate;
     OSString*      mFilePath;
-    IOTimerEventSource* mTimer;
+    IONotifier* disk_notifier_start_;
```
Sir, I'm afraid that now I have to sleep (3:00 AM here), I'll continue tomorrow.

Micky1979 commented:
Sir, I'm afraid that now I have to sleep (3:00 AM here), I'll continue tomorrow, good night

Micky1979 committed:
removing xcuserdata

Micky1979 committed:
removed mTimer local declaration

removed mTimer local declaration that hide already declared instance
variable.

Micky1979 commented:
A question. In kernel_patcher.c should not be 
`parse_mach(kernelData, NULL, NULL, &add_symbol, &section_handler);`
this way:
`parse_mach(kernelData, NULL, &add_symbol, (void *)&section_handler);`
or you have different macho parser with different args?

EDIT: solved

Micky1979 committed:
Buffer length fixed

Fixed a bug (introduced in commit 6dd7100) calculating the plist size
during write operation.

Micky1979 committed:
Don't sync FileNVRAM settings

Apply FileNVRAM  settings, but not synchronize them on nvram.plist

Micky1979 committed:
Register the Service if the data from the bootloader is available.

If /chosen/nvram is populated then register the service. mSafeToSync is
set to true only waiting for the filesystem (or for the early init).
Added a check in registerNVRAM() that return if already initialized.

Micky1979 committed:
Added entitlement check for SIP settings from Pike R. Alpha's fork

Sistem Integrity Protection will deny write operation to the nvram.
Added a check for kernel version greater or equal to 15. Code from Pike
R. Alpha’s fork.

Micky1979 committed:
Default save path customizable at compile time

Added the possibility to define a default path at compile time.
If USE_ROOT_DIR is definenide the plist will be saved to ./nvram.plist
instead of inside the Extra folder.
Chameleon can be installed in the ESP and so it’s Extra directory not
reachable by FileNVRAM.

Micky1979 committed:
Added a secondary method to load FileNVRAM.kext

Loading FileNVRAM w/o using the mkext archive.
FileNVRAMInfo.plist.h and FileNVRAM.binary.h are in order the
Info.plist and the FileNVRAM executable. (Tested and working)

Micky1979 committed:
Hook also if no nvram.plist is found.

Even if no nvram.plist is found register the call backs. This way
command line args and boot options can be handled.

Micky1979 committed:
Added some message for the boot-log

Messages for the bootlog indicating success/failures loading the plist
file.

Micky1979 committed:
Using strlen instead of sizeof for string length calculations

Using strlen instead of sizeof for string length calculations for
readability

Micky1979 committed:
Removed unused arg

Micky1979 committed:
rework of scanforNVRAM()

scanforNVRAM() function reworked, added compatibility in the case
FileNVRAM.kext saves the plist in “/“:
the module will anyway look in Extra/Extensions/ for the nvram file
ensuring that users will have their data loaded switching to this new
way.

Micky1979 committed:
is_module_loaded require only one arg

is_module_loaded in Chameleon and braches require only one arg.

Micky1979 committed:
adapted kernel's patcher section_handler() and parse_mach() to load with acual chameleon

adapted kernel's patcher section_handler() and parse_mach() to load
with acual chameleon trunk and branches

Micky1979 committed:
forgetfulness in previous commit

Micky1979 committed:
Fixed __KLD address in kernel patcher

__KLD address converted to 32bit.
In OS X Lion the module is working but in newer OSes this fix is not
enough to patch readBooterExtensions()/ readPrelinkedExtensions() even
if this address appear to be right. Should we look at the xnu source
because something has changed.

Micky1979 committed:
Making module compilable inside chameleon (part 1)

header paths chameleon compatible

Micky1979 committed:
Making module compilable inside chameleon (part 2)

moved files in parent folder, deleting unused folder

Micky1979 committed:
Making module compilable inside chameleon (part 3, the last)

The module is now compilable inside chameleon/i386/modules/.
The new makefile is customizable to use the mkext (USE_MKEXT = 1), or a
kext binary+Info.plist (USE_MKEXT = 0).
USE_ROOT_DIR = 0 set the usual save path at “/Extra/nvram.plist” folder
while USE_ROOT_DIR = 1 set the deault path to “/.nvram.plist”.
In case of building the mkext osascript will allow you to run as root
temporarily (a popup window will show up asking for the password).
Added a secondary FileNVRAM_XCODE5.xcodeproj to build the kext with
newer clang based on its version, aquired by the “buildkext” script.
Users with the old module and with nvram file inside /Extra, will have
no troubles using this version in case of a new file path.
All tested and working at insanelymac.com by many users.

Micky1979 commented:
Hi meklort. Made raical changes and I'm ready to discuss them. All tested and working by many users at insanelymac.com.

meklort commented:
I haven't had too much time to look inot this. I did start to get FileNVRAM running on my mac so I can test the module changes, but I ran into an issue with IONVMeFamily depending on AppleEFINVRAM... which will end up causing problems. I have a fix for it, but I need to setup a second boot os so that I can test it safely.

I'm also do not like the CSR/SIP change. I setup a branch that implements it the same way that Apple does, however that uses some private symbols (that can be linked to), but that does require two separate kext to work properly. There is another, safer, (and more future proof) way to do this, however it has it's own drawbacks. Specifically, it means I easily log if certain nvram variables exist, and may complicate file writing. 

Finally, please note that the module system interface is different form trunk beeches I have a separate repo with some bugfixes for it. Specifically with to the way memory is initialized. Presently there is a bug in the module system that results in bss memory being initialized wrong, as well as there being the possible that ram isn't allocated for the module. I'll see if I can update the patch to have a new api instead of changing the arguments to existing apis, however I don't know when I'll get around to it. I'd prefer to submit the changes to the forge repo instead of switching things here.

So.. I'd like to look at fixing the CSR code and getting my test system setup before going through this pull request extensively.

If, however, you were willing to create a couple of separate pull requests, i'd be happy to review a pull request with the following changes removed:
- Module interface changes removed
- Build system changes removed
- CSR changes removed.

Micky1979 commented:
Hi Sir, no problems, what I want is this module in a full working state. I'll be happy to takes a look to your  chameleon-repo (I cannot find it..) and import those changes here:https://github.com/Micky1979/Enoch or fork it, if isn't a problem. Also I like to stand on your work line, so yes I'll do what you ask and making separate pull requests with things removed as your wish (maybe from a secondary branch), with no rush.
I'm waiting for a new reponse about your chameleon-repo then I'll close this pull request.

meklort commented:
My chameleon changes were done over here:
https://public.xzenue.com/diffusion/C/history/master/

Please note that at lot of the changes were me playing around with it. I essentially have chameleon 2 from forge running in a module, which is what caused me to find the module system bug.
The other changes were some basic tests and code coverage support for chameleon when in a VM or with a serial port. That allowed me to easily create tests for regressions and to ensure that those tests covered the code I was expecting them to cover.

Micky1979 commented:
Wow, chameleon is a module o_O 
Changing the XNU string  (.patched) did not cause problems in WMWare?
..ok ok closing this pullrequest I have something to study, thanks!
